### PR TITLE
Instrument SDK code with logging statements

### PR DESCRIPTION
**Pull request by maliroteh**
@smcnulty-sfdc @wpeter-sfdc @kchitalia

This work is to instrument the SDK code with logging calls. Please take a look and give your feedback. Particularly please let me know if there are any customer sensitive data that I'm currently logging which should not be logged. Thanks

### CODE REVIEW

**Comment by smcnulty-sfdc** [comment]
1) There should be some level of control over how much is logged; right now it's extremely verbose and there is no control over if some of it should only exist in debug builds vs release. For example you log in one method that WipePincode is being called, and then in WipePincode you have an additional 4 log statements.  A better design would be to have a log statement at the start of each method saying it is called, instead of adding a log statement before every instance of the method when it is being used. 

2) I question all of the extra logging being added to the SDK in general.  I support logging when exceptions are being caught, and perhaps adding a log statement for the entry of a method that is considered critical, but not adding a log statement for every line of code.

I can't approve this as is.

**Comment by maliroteh** [comment]
Sean, I updated the call. Most of them are now being logged at Verbose level which means that only our debug build will pick these up (will be ignored on non-debug builds).

Exceptions are still logged at Critical level regardless of the build flavor.

File: SalesforceSDK/Salesforce.SDK.Core/Source/App/SalesforceApplication.cs [hunk 1]
```diff
@@ -168,12 +168,18 @@ protected void CreateClientManager(bool reset)
         {
             try
             {
+                SendToCustomLogger("SalesforceApplication.RefreshToken - calling OAuth2.RefreshAuthToken", LoggingLevel.Verbose);
```
Comment (smcnulty-sfdc) [inline SalesforceSDK/Salesforce.SDK.Core/Source/App/SalesforceApplication.cs]: This should be inside RefreshAuthToken, not here.

File: SalesforceSDK/Salesforce.SDK.Core/Source/App/SalesforceApplication.cs [hunk 2]
```diff
@@ -168,12 +168,18 @@ protected void CreateClientManager(bool reset)
...
                 await OAuth2.RefreshAuthToken(AccountManager.GetAccount());
+
+                SendToCustomLogger("SalesforceApplication.RefreshToken - calling OAuth2.RefreshCookies", LoggingLevel.Verbose);
```
Comment (smcnulty-sfdc) [inline SalesforceSDK/Salesforce.SDK.Core/Source/App/SalesforceApplication.cs]: This logline should be inside of RefreshCookies.

Comment (smcnulty-sfdc) [inline SalesforceSDK/Salesforce.SDK.Core/Source/App/SalesforceApplication.cs]: RefreshCookies log should be in the RefreshCookies method.

File: SalesforceSDK/Salesforce.SDK.Core/Source/Auth/AccountManager.cs [hunk 3]
```diff
@@ -48,11 +49,18 @@ public static void DeleteAccount()
...
+            SalesforceApplication.SendToCustomLogger("AccountManager.DeleteAccount - account deleted", LoggingLevel.Verbose);
         }
 
         public static Dictionary<string, Account> GetAccounts()
         {
-            return AuthStorageHelper.GetAuthStorageHelper().RetrievePersistedCredentials();
+            Dictionary<string, Account> accounts = AuthStorageHelper.GetAuthStorageHelper().RetrievePersistedCredentials();
+
+            SalesforceApplication.SendToCustomLogger(
+                string.Format("AccountManager.GetAccounts - Done. Total number of accounts retrieved = {0}",
```
Comment (smcnulty-sfdc) [inline SalesforceSDK/Salesforce.SDK.Core/Source/Auth/AccountManager.cs]: This should occur in RetrievePersistedCredentials.

File: SalesforceSDK/Salesforce.SDK.Core/Source/Auth/AccountManager.cs [hunk 4]
```diff
@@ -61,38 +69,52 @@ public static void DeleteAccount()
...
+            Account account = AuthStorageHelper.GetAuthStorageHelper().RetrieveCurrentAccount();
+            SalesforceApplication.SendToCustomLogger("AccountManager.GetAccount - retrieved current account", LoggingLevel.Verbose);
+            return account;
         }
 
         public static async Task<bool> SwitchToAccount(Account account)
         {
             if (account != null && account.UserId != null)
             {
+                SalesforceApplication.SendToCustomLogger("AccountManager.SwitchToAccount - save pin timer", LoggingLevel.Verbose);
```
Comment (smcnulty-sfdc) [inline SalesforceSDK/Salesforce.SDK.Core/Source/Auth/AccountManager.cs]: This should be in SavePinTimer();

File: SalesforceSDK/Salesforce.SDK.Core/Source/Auth/AccountManager.cs [hunk 5]
```diff
@@ -61,38 +69,52 @@ public static void DeleteAccount()
...
                 PincodeManager.SavePinTimer();
+                SalesforceApplication.SendToCustomLogger("AccountManager.SwitchToAccount - persist credentials", LoggingLevel.Verbose);
```
Comment (smcnulty-sfdc) [inline SalesforceSDK/Salesforce.SDK.Core/Source/Auth/AccountManager.cs]: This should be in PersistCredentials.

File: SalesforceSDK/Salesforce.SDK.Core/Source/Auth/AccountManager.cs [hunk 6]
```diff
@@ -61,38 +69,52 @@ public static void DeleteAccount()
...
                 AuthStorageHelper.GetAuthStorageHelper().PersistCredentials(account);
+                SalesforceApplication.SendToCustomLogger("AccountManager.SwitchToAccount - peek client", LoggingLevel.Verbose);
```
Comment (smcnulty-sfdc) [inline SalesforceSDK/Salesforce.SDK.Core/Source/Auth/AccountManager.cs]: This should be in PeekRestClient();

File: SalesforceSDK/Salesforce.SDK.Core/Source/Auth/AccountManager.cs [hunk 7]
```diff
@@ -61,38 +69,52 @@ public static void DeleteAccount()
...
                 RestClient client = SalesforceApplication.GlobalClientManager.PeekRestClient();
                 if (client != null)
                 {
+                    SalesforceApplication.SendToCustomLogger("AccountManager.SwitchToAccount - clear cookies", LoggingLevel.Verbose);
                     OAuth2.ClearCookies(account.GetLoginOptions());
+                    SalesforceApplication.SendToCustomLogger("AccountManager.SwitchToAccount - call identity service", LoggingLevel.Verbose);
```
Comment (smcnulty-sfdc) [inline SalesforceSDK/Salesforce.SDK.Core/Source/Auth/AccountManager.cs]: This should be in CallIdentityService.

File: SalesforceSDK/Salesforce.SDK.Core/Source/Auth/AccountManager.cs [hunk 8]
```diff
@@ -61,38 +69,52 @@ public static void DeleteAccount()
...
                     IdentityResponse identity = await OAuth2.CallIdentityService(account.IdentityUrl, client);
                     if (identity != null)
                     {
                         account.UserId = identity.UserId;
                         account.UserName = identity.UserName;
                         account.Policy = identity.MobilePolicy;
+                        SalesforceApplication.SendToCustomLogger("AccountManager.SwitchToAccount - persist credentials", LoggingLevel.Verbose);
                         AuthStorageHelper.GetAuthStorageHelper().PersistCredentials(account);
```
Comment (smcnulty-sfdc) [inline SalesforceSDK/Salesforce.SDK.Core/Source/Auth/AccountManager.cs]: This should be in PersistCredentials(account);

File: SalesforceSDK/Salesforce.SDK.Core/Source/Auth/AccountManager.cs [hunk 9]
```diff
@@ -61,38 +69,52 @@ public static void DeleteAccount()
...
                     }
+                    SalesforceApplication.SendToCustomLogger("AccountManager.SwitchToAccount - refresh cookies", LoggingLevel.Verbose);
```
Comment (smcnulty-sfdc) [inline SalesforceSDK/Salesforce.SDK.Core/Source/Auth/AccountManager.cs]: RefreshCookies() ... as commented earlier, if the log point was put into that method, you wouldn't need it again here.

File: SalesforceSDK/Salesforce.SDK.Phone/Source/Auth/AuthHelper.cs [hunk 10]
```diff
@@ -64,12 +64,15 @@ public sealed class AuthHelper : IAuthHelper
             var frame = Window.Current.Content as Frame;
             if (account.Policy != null && (!PincodeManager.IsPincodeSet() || PincodeManager.IsPincodeRequired()))
             {
+                SalesforceApplication.SendToCustomLogger("AuthHelper.EndLoginFlow - Launching Pincode Screen", LoggingLevel.Information);
```
Comment (smcnulty-sfdc) [inline SalesforceSDK/Salesforce.SDK.Phone/Source/Auth/AuthHelper.cs]: Should be in LaunchPincodeScreen().

File: SalesforceSDK/Salesforce.SDK.Store/Source/Auth/AuthHelper.cs [hunk 11]
```diff
@@ -66,12 +65,15 @@ public sealed class AuthHelper : IAuthHelper
             Account account = await AccountManager.CreateNewAccount(loginOptions, authResponse);
             if (account.Policy != null && (!PincodeManager.IsPincodeSet() || PincodeManager.IsPincodeRequired()))
             {
+                SalesforceApplication.SendToCustomLogger("AuthHelper.EndLoginFlow - Launching Pincode Screen", LoggingLevel.Information);
```
Comment (smcnulty-sfdc) [inline SalesforceSDK/Salesforce.SDK.Store/Source/Auth/AuthHelper.cs]: Should be in LaunchPincodeScreen(), won't need to duplicate this call then.

**Pull request merged**